### PR TITLE
Hotkey closing for several windows implemented (using global hotkeys)

### PR DESCRIPTION
**Pull request by Nystul-the-Magician**

### CODE REVIEW

File: Assets/Scripts/Game/UserInterfaceWindows/DaggerfallCharacterSheetWindow.cs [hunk 1]
```diff
@@ -162,6 +162,23 @@ protected override void Setup()
             characterPortrait.Refresh();
         }
 
+        public override void Update()
+        {
+            base.Update();
+
+            // check for global character sheet hotkey and close window if pressed
+            KeyCode[] hotKeyCodesCharacterSheet = InputManager.Instance.GetBindings(InputManager.Actions.CharacterSheet);
+            for (int i = 0; i < hotKeyCodesCharacterSheet.Length; i++)
```
[the remark this answers is not in the document]
Comment (Interkarma) [reply Assets/Scripts/Game/UserInterfaceWindows/DaggerfallCharacterSheetWindow.cs]: I'm not sure why a loop here is needed at all? Just sample if the player has pressed that action key:

    InputManager.Instance.HasAction(InputManager.Actions.CharacterSheet))

**Comment by Interkarma** [comment]
Nystul, I'm sorry to be critical but this is a bit messier than I envisioned. I was expecting just a couple of simple edits along the lines of: if HasAction(openAction) then CloseWindow()).

I also thought we were only going to do this from the top-level window. I'd prefer not to involve any popups or second-level windows at this time.

I'm personally not a fan of the toggle paradigm (I actually despise it), but I was willing to make a couple of simple compromises as a couple of people indicated they would like this. I would prefer it to be much more low-touch than this.

File: Assets/Scripts/Game/UserInterfaceWindows/DaggerfallCharacterSheetWindow.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Nystul-the-Magician) [reply Assets/Scripts/Game/UserInterfaceWindows/DaggerfallCharacterSheetWindow.cs]: reason is duplicate key bindings

**Comment by Nystul-the-Magician** [comment]
don't understand the line "if HasAction(openAction) then CloseWindow()".
this seems strange to me - something "if openAction then close" seems contradictious

**Comment by Nystul-the-Magician** [comment]
when you say "I also thought we were only going to do this from the top-level window. I'd prefer not to involve any popups or second-level windows at this time." do you only mean the status info message?

**Comment by Nystul-the-Magician** [comment]
btw, I want to point out that the status info pop up does show the health messagebox when enter is pressed in vanilla. So my change fixes this, because it was not working as this before - pressing enter closed the status popup without showing the messagebox

**Comment by Interkarma** [comment]
> don't understand the line "if HasAction(openAction) then CloseWindow()".
this seems strange to me - something "if openAction then close" seems contradictious

The openAction is whatever action key is used to open that window. For example character sheet window, all you should need is the following:

    if (InputManager.Instance.HasAction(InputManager.Actions.CharacterSheet))
        CloseWindow();

That's it. :)

**Comment by Interkarma** [comment]
Thank you. If input system isn't sampling these changes while game is paused, that's something I'll try and fix later.

Like I said in the original forum post, these things are much easier to ask for than to implement. I personally think this is a very low-value change and would prefer everyone focus on gameplay features.

I mean, the Escape key works just bloody fine. :)

**Comment by Nystul-the-Magician** [comment]
I disagree a bit here - I will give an example. If you want to walk to an specific building you have seen on the automap you constantly want to make some walk progress open automap to check and close it again. You don't want to press automap hotkey and esc alternating all the time, it is MUCH more convenient to use same key for open and close

**Comment by Nystul-the-Magician** [comment]
I will close for now and see if I can make it simpler

**Comment by Interkarma** [comment]
> I disagree a bit here - I will give an example. If you want to walk to an specific building you have seen on the automap you constantly want to make some walk progress open automap to check and close it again. You don't want to press automap hotkey and esc alternating all the time, it is MUCH more convenient to use same key for open and close

That's fine, and I'm happy for other people to have this. I'm just an Escape purist personally. :)

I've just posted an example with commit comments on how I'd prefer to implement this for windows like Inventory, Character, Spellbook (I'll do this one if you like), and Automap.

Using this on Info popup is fine as long as the key will show health status then close. I didn't test your earlier changes, so it that's how it worked then all is good here too. :)

**Comment by Nystul-the-Magician** [comment]
ok thanks. will merge and learn ;)

File: Assets/Scripts/Game/DaggerfallUI.cs [hunk 2]
```diff
@@ -1157,10 +1157,12 @@ void DisplayStatusInfo()
         {
             // Setup status info as the first message box.
             DaggerfallMessageBox statusBox = new DaggerfallMessageBox(Instance.uiManager, Instance.uiManager.TopWindow);
+            statusBox.KeyCodesThatCloseCurrentMessageBox = InputManager.Instance.GetBindings(InputManager.Actions.Status);
```
[the remark this answers is not in the document]
Comment (Nystul-the-Magician) [reply Assets/Scripts/Game/DaggerfallUI.cs]: sry but what you want is exactly what I implemented there. It closed the first info box and showed the health info box.

**Pull request closed without merging**